packages/cpp/src/browser/cpp-build-configurations.ts

```diff
@@ -189,12 +115,28 @@ export class CppBuildConfigurationChanger implements QuickOpenModel {
     }
 
     open() {
+        const configs = this.cppBuildConfigurations.getValidConfigs();
         this.quickOpenService.open(this, {
-            placeholder: 'Choose a build configuration...',
+            placeholder: (configs.length) ? 'Choose a build configuration...' : 'No build configurations present',
             fuzzyMatchLabel: true,
             fuzzyMatchDescription: true,
         });
     }
+
+    /** Create a new build configuration with placeholder values.  */
+    async createConfig(): Promise<void> {
+        const resource = await this.getPreferenceProvider(PreferenceScope.Workspace).source;
```

Comment:
there is `getUri` method, why to expose internals?

Reply:
@akosyakov I'm not certain as to which `getUri` I should be using to obtain the uri of the workspace preferences. The only method I could find was from the `WorkspacePreferenceProvider` from my initial commit and it was similar to what the `@theia/preferences` uses as well.

https://github.com/theia-ide/theia/blob/f3fb3095ce71e2a0f570d3980f8bce4d704ed8cd/packages/preferences/src/browser/preferences-contribution.ts#L73

Reply:
`PreferenceProvider` are not supposed to depend on `Resource`, there can be other implementations, so its should be reverted for sure. 

How about we introduce a command to open workspace preferences which will be implemented by `@theia/preferences` extension and you call it from here?

Also please if you change some APIs assign people who designed this APIs for reviews.

Reply:
For example `PreferenceProvider` based on the schema: https://github.com/theia-ide/theia/pull/2701/files#diff-efc63fa928532aa0ab10afc95417b026R53

Reply:
Thank you and yes I will be reverting the middle commit for sure and use a different approach based on your feedback. Sorry about not adding the people who designed the API for review, changes were initially self-contained within `@theia/cpp` and quickly grew.

Reply:
@akosyakov the issue here is that in order to open the edited preference file, we need to get its URI, and if we solely use a `PreferenceProvider`, there is no way to guess where it got its preferences from:

https://github.com/theia-ide/theia/blob/master/packages/core/src/browser/preferences/preference-provider.ts#L46-L52

We have get/set but no way to query the provider in order to get something representing its source of preferences.

Same goes with the recently added preference widget:

https://github.com/theia-ide/theia/blob/master/packages/preferences/src/browser/preferences-contribution.ts#L40-L41

In the preferences we also kindof refer directly to the implementation detail for the user and workspace preferences. I'm not fond of it either, but it works so I'm ok.

---

In short: How should we fetch the URI of the file to open in the editor when editing the preferences?

Reply:
I suggest that we move `Open Preferences` command declaration to core extension and keep its handler in preferences extension.

From here, you do something like:
```ts
commands.executeCommand(PREFERENCES_COMMAND.OPEN.id, {
    scope: PreferenceScope.WORKSPACE 
});
````